sky/engine/tonic/dart_args.h

```diff
@@ -0,0 +1,107 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef SKY_ENGINE_TONIC_DART_ARGS_H_
+#define SKY_ENGINE_TONIC_DART_ARGS_H_
+
+#include <type_traits>
+
+#include "sky/engine/tonic/dart_converter.h"
+#include "sky/engine/tonic/dart_wrappable.h"
+#include "dart/runtime/include/dart_api.h"
+
+namespace blink {
+
+class DartArgs {
+ public:
+  explicit DartArgs(Dart_NativeArguments args) : args_(args), index_(1) { }
+
+  template<typename T>
+  T GetNext(bool* had_exception) {
+    Dart_Handle exception = nullptr;
+    T arg = DartConverter<T>::FromArguments(args_, index_++, exception);
+    *had_exception = static_cast<bool>(exception);
+    if (exception)
+      Dart_ThrowException(exception);
+    return arg;
+  }
+
+  Dart_NativeArguments args() const { return args_; }
+
+ private:
+  Dart_NativeArguments args_;
+  int index_;
+
+  DISALLOW_COPY_AND_ASSIGN(DartArgs);
+};
+
+template<typename ArgType>
+struct DartArgHolder {
+  using ValueType = typename std::remove_const<
+    typename std::remove_reference<ArgType>::type
+  >::type;
+
+  ValueType value;
+  bool had_exception;
+
+  DartArgHolder(DartArgs* it) : value(it->GetNext<ValueType>(&had_exception)) {
+  }
+};
+
+template <typename T>
+class DartDecoder {
+};
+
+template <typename C, typename ResultType, typename... ArgTypes>
+struct DartDecoder<ResultType (C::*)(ArgTypes...)>
+    : public DartArgHolder<ArgTypes>... {
+  using FunctionPtr = ResultType (C::*)(ArgTypes...);
+
+  DartArgs* it_;
+
+  DartDecoder(DartArgs* it)
+      : DartArgHolder<ArgTypes>(it)..., it_(it) { }
+
+  bool had_exception() {
+    return Or(DartArgHolder<ArgTypes>::had_exception...);
+  }
+
+  ResultType Dispatch(FunctionPtr func) {
+    return (GetReceiver<C>(it_->args())->*func)(DartArgHolder<ArgTypes>::value...);
+  }
+
+ private:
+  static bool Or() { return false; }
+  template <typename... T>
+  static bool Or(bool arg1, T... args) {
+    return arg1 || Or(args...);
+  }
+};
+
+template<typename T>
+void DartReturn(T result, Dart_NativeArguments args) {
+  DartConverter<T>::SetReturnValue(args, result);
+}
+
+template<typename Sig>
+void DartCall(Sig func, Dart_NativeArguments args) {
+  DartArgs it(args);
+  DartDecoder<Sig> decoder(&it);
```

Comment:
When you call this function, you pass a member function as the first argument.  For example, `&SceneBuilder::pushClipRRect`.  That has a type roughly like `void (SceneBuilder::*)(const RRect&, const Rect&)`.  The compiler decides `Sig` ought to have that type as its value in order to make the code build (template deduction).

Now, when you give that type to `DartDecoder<xxx>`, it tries to instantiate that template.  It has several choices, for example, the generic `DartDecoder` that matches any type and has no code or the more elaborate `DartDecoder` with the following declaration:

```
template <typename C, typename ResultType, typename... ArgTypes>
+struct DartDecoder<ResultType (C::*)(ArgTypes...)>
```

This second `DartDecoder` is a partial specialization of the first, so the compiler sees if the thing inside the `<xxx>` has the same structure as `Sig`.  In this case it does, so it tries to work backwards to fill in the template parameters.  It succeeds, finding `C` is `SceneBuilder`, `ResultType` is `void`, and that there are two `ArgTypes`, `const Rect&` and `const RRect&`.  It then instantiates the rest of the template with those values and sees if it compiles.  In this case it does, so we get the specialization we wanted.